DW-727: Add troubleshooting info of the confluence page (DOCS-12701) (#87)
mulesoft/docs-dataweave: modules/ROOT/pages/dataweave-troubleshoot.adoc

+
`-M-Dcom.mulesoft.dw.dump_folder=<path_to_folder>`
+

> The default directory is what is set in the `java.io.tmpdir` property.

Reviewer: ```suggestion
The default directory is set in the `java.io.tmpdir` property.
```

You can configure the maximum stack size by using the property `com.mulesoft.dw.stacksize`.

=== No space left on device

> In order to handle large payloads, DataWeave generates some data which is handled in memory unless it exceeds a certain configurable limit.

Reviewer: ```suggestion
To handle large payloads, DataWeave generates data that is handled in memory unless the payload exceeds a configurable limit.
```

> When that happens the data is stored on disk as output, input, and buffer files in a temporary directory. See xref:dataweave-memory-management.adoc[DataWeave Memory Management] for further details.

Reviewer: ```suggestion
If the payload exceeds the limit, the data is stored on disk as output, input, and buffer files in a temporary directory. See xref:dataweave-memory-management.adoc[DataWeave Memory Management] for further details.
```

> Once the streams referencing those files are closed the files are released. The closure normally happens when flows complete their execution, so long-running and particularly concurrent executions could involve having many of these files living at once.

Reviewer: The second sentence repeats some previous content.

Reviewer: ```suggestion
When the streams referencing those files are closed, the files are released. Closure normally occurs when flows complete their execution, so buffer files can be used within the temporary file during long-running and concurrent executions.
```

> To avoid the exception `No space left on device` more resources are needed to be able to run the app, or the app needs to be reviewed to reduce its resource consumption.

Reviewer: ```suggestion
To avoid the exception `No space left on device`, you must provide more resources to run the application or reduce its resource consumption.
```

> Sometimes bugs can lead to them not being removed even after the flow has completed.

Reviewer: ```suggestion
Sometimes, bugs can prevent release of the files after the flow has completed.
```

Author: Added `Sometimes bugs prevent the release of the files even after the flow has completed.`
Because there are not related to reducing resource consumption or providing more resources.

Reviewer: Yes, I was thinking that if I am trying to troubleshoot this issue, my steps would be to try to increase resources or reduce resource consumption, but if that didn't work, then I would assume that there is a bug to report.

> If this case isn't fixed in the latest update, there’s no way to work around the issue, and this can be reported to the Mulesoft support team.

Reviewer: ```suggestion
Such cases require an update to the code. If the latest Mule release does not fix this issue, please report the issue to the MuleSoft support team.
```

Author: Added `If the latest Mule release does not fix this issue, please report the issue to the MuleSoft support team.`

=== Closed Streams

> The input data that makes its way to DataWeave is usually a stream handled by the Mule Runtime: they will add auto-closing and repeatable capabilities by generating “cursors” over them.

Reviewer: ```suggestion
Input data that reaches DataWeave is usually a stream handled by Mule, which adds auto-closing and repeatable capabilities by generating “cursors” over the data.
```

Author: Added `"cursors" over the stream.`

> Sometimes these streams are closed prematurely by the time they reach a DataWeave script, and it can be hard to understand who has closed it.

Reviewer: ```suggestion
Sometimes a stream closes prematurely, before it reaches the DataWeave script, and it can be hard to understand what closed it.
```

> In these circumstances, the script will fail with an error similar to: "Cannot open a new cursor on a closed stream".

Reviewer: ```suggestion
In these circumstances, the script fails with an error similar to `Cannot open a new cursor on a closed stream`.
```

> If this case isn't fixed in the latest update, there’s no way to work around the issue, but there is a way to determine which component is closing the stream prematurely, so the issue can be properly reported to the Mulesoft support team.

Reviewer: ```suggestion
If the issue occurs in the latest Mule update. please report it to the MuleSoft support team. 

Although is not possible to work around this issue, you can determine which component is closing the stream prematurely.
```

Author: Added the next suggestion

> If this case isn't fixed in the latest update, there’s no way to work around the issue, but there is a way to determine which component is closing the stream prematurely, so the issue can be properly reported to the Mulesoft support team.

Reviewer: ```suggestion
If this error occurs in latest update, please report it the MuleSoft support team. It is not possible to work around the issue, but you can use the `com.mulesoft.dw.track.cursor.close` system property to determine which component closed the stream prematurely.
```

> Using the `com.mulesoft.dw.track.cursor.close` system property, the error will start showing the stack trace from the moment the stream was closed and therefore point to whoever triggered the closing.

Reviewer: ```suggestion
With the property set, the error shows the stack trace from the moment that the stream was closed, which points to the component that triggered the closure.
```

You can configure the maximum stack size by using the property `com.mulesoft.dw.stacksize`.

=== No space left on device
In order to handle large payloads, DataWeave generates some data which is handled in memory unless it exceeds a certain configurable limit.
When that happens the data is stored on disk as output, input, and buffer files in a temporary directory. See xref:dataweave-memory-management.adoc[DataWeave Memory Management] for further details.

> To avoid the exception `No space left on device` more resource is needed to been able to run the app, or the app needs to be reviewed to reduce its resource consumption.

Reviewer 2: `more resources are needed to be able`

> Normally, once the streams referencing those files are closed the files are released. In case that these types of files remains in the temporal folder even after the flow has completed, this could mean that the streams remain unclosed and cause leaks.

Reviewer 2: I would add a note saying that the closure normally happens when flows complete their execution, so long-running and particularly concurrent executions could involve having many of these files living at once. Then the explanation that sometimes bugs can lead to them not being removed and the note for support.

----

Making the output explicit solves the issue: `output application/java --- payload splitBy ' '`

== DataWeave Warnings

> The following are some common DataWeave warnings that you can find in your dump file, along with some details to address these warnings.

Reviewer: ```suggestion
You can address many common DataWeave warnings that appear in your dump file.
```

Author: Added this `The following are some common DataWeave warnings that appear in your dump file, along with some details to address these warnings.` That "DW Warnings" was already written, I just moved it to the end of the file.

Reviewer: Just to address some style issues with this paragraph by cutting out unnecessary words.

=== End of Input Was Reached

> This warning shows because DataWeave treats the Unicode non-character `U+FFFF` as a special character that indicates the end of input.

Reviewer: ```suggestion
This warning appears because DataWeave treats the Unicode non-character `U+FFFF` as a special character that indicates the end of input.
```
